router/batchrouter/asyncdestinationmanager/marketo-bulk-upload/utils.go

```diff
@@ -0,0 +1,278 @@
+	"fmt"
+	"io"
+	"mime/multipart"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/rudderlabs/rudder-server/router/batchrouter/asyncdestinationmanager/common"
+)
+
+// Docs: https://experienceleague.adobe.com/en/docs/marketo-developer/marketo/rest/error-codes
+func categorizeMarketoError(errorCode string) (status string, message string) {
+	// Convert string error code to integer
+	code, err := strconv.Atoi(errorCode)
+	if err != nil {
+		return "Retryable", "Unknown"
+	}
+
+	switch {
+
+	// Specific retryable errors mentioned in the document
+	case code == 502:
+		return "Retryable", "Bad Gateway"
+	case code == 604:
+		return "Retryable", "Request time-out"
+	case code == 608:
+		return "Retryable", "API Temporarily Unavailable"
+	case code == 611:
+		return "Retryable", "System error"
+	case code == 614:
+		return "Retryable", "Invalid Subscription"
+	case code == 713:
+		return "Retryable", "Transient Error"
+	case code == 719:
+		return "Retryable", "Lock wait timeout exception"
+
+	// 5XX errors are generally retryable
+	case code >= 500 && code < 600:
+		return "Retryable", "Internal Server Error"
+
+	// Rate limiting and quota errors
+	case code == 606:
+		return "Throttle", "Max rate limit exceeded"
+	case code == 607:
+		return "Throttle", "Daily quota reached"
+	case code == 615:
+		return "Throttle", "Concurrent access limit reached"
+	case code == 1029:
+		return "Throttle", "Too many jobs in queue or Export daily quota exceeded"
+
+	// 4XX errors are generally not retryable (abortable)
+	case code >= 400 && code < 500:
+		return "Abortable", "Unknown"
+
+	case code == 601:
+		return "RefreshToken", "Access token invalid"
+	case code == 602:
+		return "RefreshToken", "Access token expired"
+
+	// Specific abortable errors
+	case code == 603:
+		return "Abortable", "Access denied"
+	case code == 605:
+		return "Abortable", "HTTP Method not supported"
+	case code == 609:
+		return "Abortable", "Invalid JSON"
+	case code == 610:
+		return "Abortable", "Requested resource not found"
+	case code == 612:
+		return "Abortable", "Invalid Content Type"
+	case code == 616:
+		return "Abortable", "Invalid subscription type"
+
+	// Add more specific error codes as needed
+
+	default:
+		return "Retryable", "Unknown"
+	}
+}
+
+func handleMarketoErrorCode(errorCode string) (int64, string, string) {
+	category, message := categorizeMarketoError(errorCode)
+	switch category {
+	case "Retryable":
+		return 500, category, message
+	case "Throttle":
+		return 429, category, message
+	case "Abortable":
+		return 400, category, message
+	case "RefreshToken":
+		return 500, category, message
+	default:
+		return 500, category, message
+	}
+}
+
+func parseMarketoResponse(marketoResponse MarketoResponse) (int64, string, string) {
+	statusCode := int64(200)
+	category := "Success"
+	errorMessage := ""
+	if !marketoResponse.Success {
+		if len(marketoResponse.Errors) > 0 {
+			errorCode := marketoResponse.Errors[0].Code
```

Comment:
Don't we need to consider other errors from errors array?

Reply:
We only get 1 error object form error array so far thats what we have seen